apps/minifront/src/state/staking.ts

```diff
@@ -162,26 +169,67 @@ export const createStakingSlice = (): SliceCreator<StakingSlice> => (set, get) =
     set(state => {
       state.staking.delegationsByAccount.set(addressIndex.account, []);
       state.staking.votingPowerByValidatorInfo = {};
+      state.staking.loading = true;
     });
 
-    for await (const delegation of getDelegationsForAccount(addressIndex)) {
-      const delegations = get().staking.delegationsByAccount.get(addressIndex.account) ?? [];
+    let delegationsToFlush: ValueView[] = [];
+    let allDelegationsLoaded = false;
 
-      const sortedDelegations = [...delegations, delegation].sort(byBalanceAndVotingPower);
+    /**
+     * Per the RPC call, we get delegations in a stream, one-by-one. If we push
+     * them to state as we receive them, React has to re-render super
+     * frequently. Rendering happens synchronously, which means that the `for`
+     * loop below has to wait until rendering is done before moving on to the
+     * next delegation. Thus, the staking page loads super slowly if we render
+     * delegations as soon as we receive them.
+     *
+     * To resolve this performance issue, we instead queue up a number of
+     * delegations (up to `DELEGATIONS_FLUSH_TO_STATE_INTERVAL`), and then flush
+     * them to state in batches.
+     */
+    const flushToState = (ignoreInterval = false) => {
+      if (allDelegationsLoaded) return;
 
-      set(state => {
-        state.staking.delegationsByAccount.set(addressIndex.account, sortedDelegations);
-      });
+      if (delegationsToFlush.length >= DELEGATIONS_FLUSH_TO_STATE_INTERVAL || ignoreInterval) {
+        const delegations = get().staking.delegationsByAccount.get(addressIndex.account) ?? [];
+
+        const sortedDelegations = [...delegations, ...delegationsToFlush].sort(
+          byBalanceAndVotingPower,
+        );
 
+        set(state => {
+          state.staking.delegationsByAccount.set(addressIndex.account, sortedDelegations);
+        });
+
+        delegationsToFlush = [];
+      }
+
+      requestAnimationFrame(() => flushToState());
+    };
+
+    let isFirstDelegation = true;
+    for await (const delegation of getDelegationsForAccount(addressIndex)) {
+      delegationsToFlush.push(delegation);
       validatorInfos.push(getValidatorInfoFromValueView(delegation));
+
+      if (isFirstDelegation) {
+        requestAnimationFrame(() => flushToState());
+        isFirstDelegation = false;
+      }
     }
 
+    // One last flush, in case the number of remaining delegations is less than
+    // `FLUSH_INTERVAL`.
+    flushToState(true);
+    allDelegationsLoaded = true;
```

Comment:
Given `allDelegationsLoaded = true;` is right before this exits, I at first was going to question whether it needs to exist at all. But given it's not set true anywhere else, does that mean we have a potential infinite loop here? flushToState() seems to run infinitely (?).

Reply:
[This line](https://github.com/penumbra-zone/web/pull/705/files#diff-7c5ec26a35f650875c894552352cecdc5e1042768e81d382bffa45e0a4dc0c1eR196) stops the loop if `allDelegationsLoaded` is true.